onnx/defs/nn/defs.cc

```diff
@@ -2190,6 +2190,183 @@ ONNX_OPERATOR_SET_SCHEMA(
         }
         )ONNX"));
 
+void col2imShapeInference(InferenceContext& ctx) {
+  propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+  // All inputs shapes are required
+  if (!hasNInputShapes(ctx, 3)) {
+    return;
+  }
+
+  // We assume image_shape has correct spatial dimensions for next validations
+  // An alternative is get the the number of spatial dimensions as an input argument
+  Dim n_input_dims;
+  unifyInputDim(ctx, 1, 0, n_input_dims);
+  std::vector<int64_t> image_shape = {};
+  const TensorProto* image_shape_data = ctx.getInputData(1);
+  if (image_shape_data) {
+    image_shape = ParseData<int64_t>(image_shape_data);
+    unifyDim(n_input_dims, image_shape.size());
+  }
+
+  std::vector<int64_t> pads = {};
+  if (getRepeatedAttribute(ctx, "pads", pads)) {
+    if (pads.size() % 2) {
+      fail_shape_inference("Attribute pads must have an even size");
+    }
+    unifyDim(n_input_dims, pads.size() / 2);
+  }
+
+  std::vector<int64_t> dilations = {};
+  if (getRepeatedAttribute(ctx, "dilations", dilations)) {
+    unifyDim(n_input_dims, dilations.size());
+  }
+
+  std::vector<int64_t> strides = {};
+  if (getRepeatedAttribute(ctx, "strides", strides)) {
+    unifyDim(n_input_dims, strides.size());
+  }
+
+  auto input_shape = ctx.getInputType(0)->tensor_type().shape();
+  if (input_shape.dim_size() != 3) {
+    fail_shape_inference("input must have rank 3.");
+  }
+
+  std::vector<int64_t> block_shape = {};
+  const TensorProto* block_shape_data = ctx.getInputData(2);
+  if (block_shape_data) {
+    block_shape = ParseData<int64_t>(block_shape_data);
+    unifyDim(n_input_dims, block_shape.size());
+  }
+  unifyInputDim(ctx, 2, 0, n_input_dims);
+
+  int block_shape_size = 0;
+  if (static_cast<int>(block_shape.size()) > 0) {
+    block_shape_size = 1;
+    for (const auto& dim : block_shape) {
+      block_shape_size *= dim;
```

Comment:
I believe this should account for dilation, if dilation is specified?

Reply:
AFAIK, for computing `C`, we don't need to account for dilation. As an example, on ORT's Col2Im, a `dilated_kernel_shape` was computed for calculating the new `image_size` when calling Im2Col, but `C` for Col2Im is still computed using regular `kernel_shape_size`

![image](https://user-images.githubusercontent.com/5469809/185209938-55911d2f-adc5-413d-9690-0fc3f0e619ec.png)

Reply:
Ok, I misunderstood dilations. I guess the kernel-block-size remains the same with or without dilations. So, this is fine, thanks!